### PR TITLE
Implement basic touch support

### PR DESCRIPTION
**Pull request by Danacus**
I've implemented basic touch (finger) input by adding another another invisible cursor to the seat. This cursor never moves as it is not needed unless we wanted to emulate a mouse pointer for surfaces that don't support touch (all modern GTK3 and Qt5 programs do support touch). Events are propagated to clients and don't interact with the cursor.

Whenever `handleTouchDown` is called, the mouse cursor will be hidden until the mouse is moved.

I also changed the `surfaceAt` function in `Cursor` by adding a `surfaceAtPosition` function such that it can be used without a `Cursor`.

### CODE REVIEW

**Comment by Leon-Plickat** [comment]
> I've implemented basic touch (finger) input by adding another another invisible cursor to the seat. 

Why? What advantages does this have over just sending touch events.

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:26]: You don't use anything from `c.zig`, so you do not need to import it.

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:28]: You also don't use anything from `util.zig`.

**Comment by Leon-Plickat** [comment]
Have you tested this with multi touch? I am pretty sure that will break the integration with the invisible cursor.

**Comment by Leon-Plickat** [comment]
Have you tested drag-and-drop between clients? For example, does this correctly work when you try to move a file from one filemanager view to another (nautilus should implement touch, if you need a test client)?

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:108]: Why hide the main mouse cursor? It has nothing to do with the touch operation.

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:92]: If this function is called and the coordinates are on an unfocused view, it should be focused by the seat. Unless I am missing something, that is currently not the case.

**Comment by Danacus** [comment]
> > I've implemented basic touch (finger) input by adding another another invisible cursor to the seat.
> 
> ~Why? What advantages does this have over just sending touch events?~
> 
> ~I don't think this patch should touch `Cursor.zig` at all.~
> 
> ~Wayland allows clients to support touch inputs directly without going through cursor emulation and all relevant clients do so. Given that adding touch support to clients is almost trivial (I have done it myself) I don't see any reason to support clients that do not implement touch.~
> 
> Nevermind, sway does it too. I still think it's a bit of a hack though.

The thing about needing a cursor for touch seems to be something odd about wlroots. I didn't want a `wlr_cursor, because we don't need one for touch`, but that's where the events are defined, so I don't think there's another way.

The reason I had to touch `Cursor.zig` is so I wouldn't have to copy the `surfaceAt` code. Perhaps this code could be moved somewhere else?

**Comment by Danacus** [comment]
> Have you tested this with multiple touchpoints?

Yes, I can zoom with 2 fingers. More than 2 touchpoints should work too.

File: river/Touch.zig
Comment (Danacus) [reply river/Touch.zig:108]: It's something that annoys me about GNOME. When switching my laptop to tablet mode, it's a bit annoying to have a cursor in the middle of my screen that I can't move in any way because I can't use the mouse. Sway and KDE do hide the mouse cursor when using touch, but I understand that it seems a bit odd at first.

**Comment by Danacus** [comment]
> Have you tested drag-and-drop between clients? For example, does this correctly work when you try to move a file from one filemanager view to another (nautilus should implement touch, if you need a test client)?

I can't really figure out what's up with drag and drop. It works on GTK apps like nautilus on sway, but not between windows. On GNOME it just doesn't work at all surprisingly. This might have something to do with sway actually moving the mouse cursor with touch (which feels like a hack). I wonder if touch drag and drop is even a thing.

File: river/Touch.zig
Comment (Danacus) [inline river/Touch.zig:106]: This should ideally only happen if the currently focused view is different from this view. However, including such statement causes problems where it's not possible to focus the surface which has the mouse pointer on it, because it thinks that view is already focused even though it's not. I'm not sure how to deal with this and my current solution is not great.

Comment (Leon-Plickat) [reply river/Touch.zig:106]: You could try to have a new `surfaceAtPosition()` function in `Touch.zig`.

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:111]: This assert is not needed. I don't see how it will ever be reached unless XWayland support is enabled, in which case it will never fail.

File: river/Touch.zig
Comment (Danacus) [reply river/Touch.zig:106]: The `surfaceAtPosition` from `Cursor` works fine and adding one in `Touch.zig` would create a lot of code duplication which I wanted to avoid. The problem here is that I'm setting focus to the current surface without checking if that surface is already focused. Adding such check would cause the aforementioned problems.

File: river/Touch.zig
Comment (Leon-Plickat) [inline river/Touch.zig:106]: `self.seat.focus(view)` will already focus the output the view is on if it is unfocused, so you do not need to do it manually here.

Comment (Leon-Plickat) [reply river/Touch.zig:106]: You probably copied from `checkFocusFollowCursor()`, right? If you like, you can add a commit removing it there as well.

Comment (Danacus) [reply river/Touch.zig:106]: I did copy it from there and I did also notice that `seat.focus` sets the output. I'll fix it in both places.

File: river/Touch.zig
Comment (Danacus) [reply river/Touch.zig:111]: This is also just here because I copied it from somewhere else and wasn't sure what it's supposed to do so I left it.

File: river/Touch.zig
Comment (Leon-Plickat) [reply river/Touch.zig:106]: Can't you just check the focus of the seat?

```zig
    if (Cursor.surfaceAtPosition(lx, ly)) |result| {
        switch (result.parent) {
            .view => |view| {
                 if (seat.focused == .view and seat.focused.view == view) {
                    self.seat.focusOutput(view.output);
                }
            }
```

Comment (Danacus) [reply river/Touch.zig:106]: I was having problems with that earlier, but now it works fine.

File: river/Touch.zig
Comment (Danacus) [reply river/Touch.zig:112]: I think the assert is just used to assert that that case is never reached. I don't think it's needed in `checkFocusFollowsCursor` either, but it's perhaps just a "coding style" thing?

File: river/Cursor.zig
Comment (Leon-Plickat) [inline river/Cursor.zig:458]: Since this is now also used elsewhere in a different context, I am not really a fan of having it in `Cursor.zig`, seems a bit leaky abstraction-wise.

File: river/Seat.zig
Comment (Leon-Plickat) [inline river/Seat.zig:450]: I quickly checked Sway again and I think they do not have a separate cursor for touch. What are the advantages of having a separate cursor instead of just using the one that is already there?

**Review by Leon-Plickat — COMMENTED** [review]
Probably last review from me.

I can't test it, but to me it looks sound overall.

Oh, and you probably want to squash all your "Touch: ..." commits. While git commits are supposed to be incremental, I'd consider implementing touch support a single such increment that needs no splitting.

File: river/Cursor.zig
Comment (Danacus) [reply river/Cursor.zig:458]: I also considered moving it somewhere else, but wasn't sure where.

File: river/Seat.zig
Comment (Leon-Plickat) [reply river/Seat.zig:450]: And, since touch operations seem to be cursor operations in wlroots, why not have them in `Cursor.zig`?

**Comment by Leon-Plickat** [comment]
Yeah, thinking about it I believe the touch handlers should be in `Cursor.zig` as well, since they are handlers for cursor events. In that case you can also keep `surfaceAtPosition` where it is now.

**Comment by Danacus** [comment]
I guess that makes sense to follow the design of wlroots even though it feels a bit odd to me that you would need a cursor for touch. In Sway they also move the cursor on touch input, but I don't think this makes a lot of sense.

File: river/Seat.zig
Comment (Danacus) [reply river/Seat.zig:450]: I was using the mouse cursor at first but changed that when you mentioned that touch is not supposed to move the cursor. However, since wlroots needs a cursor for the events, I needed to add a cursor anyway. So perhaps it makes sense to use the already existing mouse cursor since we only really need it for events.

**Comment by Danacus** [comment]
Btw, thanks for taking your time to review my PR, I really appreciate it!

File: river/Touch.zig
Comment (Leon-Plickat) [reply river/Touch.zig]:
```IRC
12:47   leon-p: ifreund: why is there an assert(build_options.xwayland) in checkFocusFollowsCursor()? If xwayland support is not enabled at compile time, that branch will never be reached anyway
14:05  ifreund: leon-p: exactly, if build_options.xwayland is false we want to mark the branch as unreachable
```

**Review by Leon-Plickat — APPROVED** [review]

File: river/Cursor.zig
Comment (ifreund) [inline river/Cursor.zig:210]: I think we would need to save whatever the current cursor image is to be able to properly restore it on the next pointer input. Cursor images may be set by clients.

Leaving this feature for a second PR would likely get this one merged more quickly.

File: river/Cursor.zig
Comment (ifreund) [inline river/Cursor.zig:523]:
```suggestion
/// Find the surface at the given position if any, and return information about that
```

File: river/Cursor.zig
Comment (ifreund) [inline river/Cursor.zig:520]:
```suggestion
    return surfaceAtPosition(self.wlr_cursor.y, self.wlr_cursor.x);
```

**Review by ifreund — CHANGES_REQUESTED** [review]
Thanks for working on this!

> I guess it makes sense to follow the design of wlroots even though it feels a bit odd to me that you would need a cursor for touch. In Sway they also move the cursor on touch input, but I don't think this makes a lot of sense.

You don't need a cursor for touch, the events of the `wlr_touch` struct may be listened to directly. One reason to go through a wlr_cursor would be if we wanted to simulate pointer input on surfaces that don't subscribe to touch events (like sway does). Nobody's asked for this feature yet and I don't even have any touch hardware though.

Another reason to go through the wlr_cursor would be to support mapping input devices to/from outputs/regions, as wlr_cursor does the required conversions for us. I imagine that this is something we will eventually want to support.

Anyhow, I think putting this code in `Cursor.zig` is fine.

File: river/Seat.zig
Comment (Danacus) [reply river/Seat.zig]: I'm not sure how to handle the destroy signal.

Comment (ifreund) [reply river/Seat.zig]: The destroy signal on `wlr.InputDevice` is the one I was talking about here.

Comment (Danacus) [reply river/Seat.zig]: Should `Seat.zig` handle this or should it be handled in a `Touch.zig` file similar to `Keyboard.zig`? Also, when a device is destroyed, should the capabilities of the seat be updated? Shouldn't this also be the case for when a keyboard is destroyed? I'm not sure how or if this gets done right now.

**Comment by theSuess** [comment]
I've locally rebased these changes onto master and can confirm that touch works great on my Surface Pro 4! Would be awesome to have this merged

**Comment by Danacus** [comment]
There are still some things that need to be finished, but I've been busy with university. I might finish this later.

**Comment by Danacus** [comment]
It's mostly just some details as discussed in the threads on this issue. Functionally everything is done, it just needs to be finished such that maintainers accept the PR.

File: river/Seat.zig
Comment (ifreund) [reply river/Seat.zig]: Hey, sorry it's taken so long to get back to you. It looks like we don't ever update the capabilities on removal for keybords/pointers currently. I think the right way to do this would be adding a `removeDevice()` function to `Seat.zig` and call that from `InputDevice.handleDestroy()` in `InputManager.zig`.

Anyhow, since it turns out we don't do this for *any* input devices currently feel free to leave out of this PR, I can implement that when I find the time to merge this if you like.

**Comment by Danacus** [comment]
This is indeed a non-trivial addition. I once attempted this, but lost time and motivation.

These are the changes I proposed for `zig-wlroots`: https://github.com/swaywm/zig-wlroots/pull/35

And a very crude attempt to get pen input working: https://github.com/Danacus/river/tree/tablet_v2

The main struggle I had was that the implementation of `tablet-unstable-v2` in `wlroots` felt a bit unintuitive and confusing to me.

**Comment by ifreund** [comment]
I've recently refactored river's input device handling to make touch support cleaner to implement. Due to that and other changes to the cursor code I decided to continue this effort in a new PR: https://github.com/riverwm/river/pull/632

Any testing would be highly appreciated, I still don't have any touch hardware.

**Comment by ifreund** [comment]
Thanks for all your work here @Danacus, this feature has been merged in #632!

**Pull request closed without merging**
